linkerd/proxy/spire-client/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+#![deny(rust_2018_idioms, clippy::disallowed_methods, clippy::disallowed_types)]
+#![forbid(unsafe_code)]
+
+mod api;
+
+pub use api::{Api, SvidUpdate};
+use linkerd_error::{Error, Result};
+use linkerd_identity::Credentials;
+use linkerd_identity::Id;
+use std::fmt::Display;
+use std::time::{Duration, UNIX_EPOCH};
+use tokio::sync::watch;
+use tower::Service;
+use tracing::error;
+
+pub struct Spire {
+    id: Id,
+}
+
+// === impl Spire ===
+
+impl Spire {
+    pub fn new(id: Id) -> Self {
+        Self { id }
+    }
+
+    pub async fn run<C, S>(self, credentials: C, mut client: S)
+    where
+        C: Credentials,
+        S: Service<(), Response = tonic::Response<watch::Receiver<SvidUpdate>>>,
+        S::Error: Into<Error> + Display,
+    {
+        match client.call(()).await {
+            Ok(rsp) => consume_updates(&self.id, rsp.into_inner(), credentials).await,
+            Err(error) => error!(%error, "could not establish SVID stream"),
+        }
+    }
+}
+
+async fn consume_updates<C>(
+    id: &Id,
+    mut updates: watch::Receiver<api::SvidUpdate>,
+    mut credentials: C,
+) where
+    C: Credentials,
+{
+    loop {
+        let svid_update = updates.borrow_and_update().clone();
+        if let Err(error) = process_svid(&mut credentials, svid_update, id) {
+            tracing::error!("Error processing SVID update: {}", error);
+        }
+
+        if let Err(error) = updates.changed().await {
+            tracing::debug!("SVID watch closed; terminating {}", error);
+            return;
+        }
```

Comment:
nit:

```suggestion
        if let Err(error) = process_svid(&mut credentials, svid_update, id) {
            tracing::error!(%error, "Error processing SVID update");
        }

        if updates.changed().await.is_err() {
            tracing::debug!("SVID watch closed; terminating");
            return;
        }
```

We generally use field interpolation for errors (rather than direct string formatting).

Also, the error type from changed() is uninteresting -- it can only indicate that all receivers were dropped.